### DISCUSSION TITLE
GfxAlignment

### DISCUSSION THREAD

**Discussion by niziul** [scripthookvdotnet/scripthookvdotnet#1282] [Ideas]
It would be possible to introduce a functionality to align graphic elements based on the safe zone, as it is normally used by the game, to the ContainerElement class?

**Comment by kagikn** [comment]
I wouldn't mind adding functionalities to modify graphic alignment configs, but we could just config them via a static class. The problem is, all the natives with the string `SCRIPT_GFX_ALIGN` modify some of the global variables for them without any isolation or no resetting (so if you set some of them in SHVDN scripts, the changes will apply even in other game scripts and ASI scripts) and there are no getter natives for them (while text style config variables reset after each text draw).

As much as it would be nice to have if we could revert graphic alignment configs after each draw for sure, we could just config them via a static class so we won't have to worry about custom getter methods or properties getting broken by some game update. I am kind of sorry to say, but I'm not convinced that it's worth adding *our custom* getter functions just to make sure we can revert previous graphic alignment settings.

Also, it may be redundant if you use the same configs for several graphic draws and bother reverting graphic alignments after each draw of them (there's no choice for text style configs on the other hand).

**Reply by kagikn** [reply]
@niziul Are you content with #1283 in scripted 2d gfx alignment? Hope the docs attached to the features will help you.

**Reply by niziul** [reply]
@kagikn did you get any results from your tests? I tested with the code below, but did not get any result. Is it necessary to modify something else to make it work correctly?

```csharp
        public Main()
        {
            var containerElement
                = new ContainerElement()
                {
                    Position
                    = new PointF(10f,
                                 10f),

                    Size
                    = new SizeF(200,
                                100),

                    Color
                    = Color.FromArgb(255,
                                     255,
                                     255,
                                     255),
                };

            //Scripted2DGfxSettings
            //        .SetAlignmentType(UIAlignment
            //                                .Top,
            //                          UIAlignment
            //                                .Center);

            Tick += (o, e) =>
            {
                if (Game.IsLoading)
                {
                    return;
                }

                Scripted2DGfxSettings
                    .SetAlignmentType(UIAlignment
                                            .Top,
                                      UIAlignment
                                            .Center);
                containerElement
                    .ScaledDraw();

            };
        }
```

**Reply by kagikn** [reply]
`Scripted2DGfxSettings.SetAlignmentType` was using an incorrect native and fixed in 4dc9705. You should set and reset the alignment like this:

```cs
Scripted2DGfxSettings.SetAlignmentType(UIAlignment.Right, UIAlignment.Center);

containerElement.ScaledDraw();

Scripted2DGfxSettings.ResetAlignment(); // You want to reset the alignment before passing the right to execute some code to other scripts, right? If you doubt it, remove this line and see what happens.
```

**Reply by niziul** [reply]
I did some tests here, and it works most of the time, but I saw that "`UIAlignment.Center`" does not work in "`ContainerElement`", and "`UIAlignment.Bottom`", "`UIAlignment.Right`" ends up not working in "`TextElement`", but "`UIAlignment.Center`" works normally. Have any of your tests given similar results?

**Reply by kagikn** [reply]
I have no idea what source says `UIAlignment.Center` does not work in `ContainerElement`, but I showed you a wrong example (corrected now). You need to put one of `UIAlignment.Left`, `UIAlignment.Right`, or `UIAlignment.Center` for the X-axis alignment (horizontal alignment) parameter to draw 2D gfx differently, and for the Y-axis alignment (vertical alignment) parameter, `UIAlignment.Top`, `UIAlignment.Bottom`, or `UIAlignment.Center`.

**Reply by niziul** [reply]
I believe it is something misconfigured in my game, because `UIAlignment.Center` ends up not working in `ContainerElement`.

**Reply by kagikn** [reply]
Okay, `UIAlignment.Center` behaved in a confusing way. No ysc scripts use that value either. I removed that value in c46c5dc. Try the code below to see how `SET_SCRIPT_GFX_ALIGN_PARAMS` works.
```cs
ContainerElement mContainer2 = new ContainerElement(new Point(0, 0), new Size(500, 20), Color.Red);
GTA.Graphics.Scripted2DGfxSettings.SetAlignmentType(GTA.Graphics.UIAlignment.Ignore, (GTA.Graphics.UIAlignment)67); // Compare with the statement where GTA.Graphics.UIAlignment.Bottom is put at the alignY param! This method calls SET_SCRIPT_GFX_ALIGN_PARAMS.
GTA.Graphics.Scripted2DGfxSettings.SetAlignmentOffsetAndSize(new PointF(0, 0f), new SizeF(0f, 0.5f));
mContainer2.Draw();
GTA.Graphics.Scripted2DGfxSettings.ResetAlignment();
```

The game seems to calculate the offset like this:
```cpp
case 66: // 'B'
    calculated_offset = offset - bottom_pos;
    break;
case 67: // 'C'
    calculated_offset = (float)((float)(offset * 2.0) - bottom_pos[1]) * 0.5;
    break;
```
Note that drawing methods in classes in `GTA.UI` are scaled by either 1280 or 720.

**Comment by kagikn** [comment]
Do you mind methods not having screen pixel adjustment like most of existing UI methods or properties do? That screen pixel adjustment (1280x720 res one) doesn't make much sense if you draw something on a render target (texture) for props... You wouldn't think the adjustment works too well for the orbital cannon map prop `xm_prop_orbital_cannon_table`.
